### PR TITLE
RAM Optimizations for M0

### PR DESCRIPTION
**Pull request by brentru**

### CODE REVIEW

File: adafruit_tinylora/adafruit_tinylora.py [hunk 1]
```diff
@@ -215,7 +199,7 @@ def send_data(self, data, data_length, frame_counter, timeout=2):
         enc_data = bytearray(data_length)
         lora_pkt = bytearray(64)
         # copy bytearray into bytearray for encryption
-        for i in range(0, data_length):
+        for i in range(data_length):
```
Comment (ladyada) [inline adafruit_tinylora/adafruit_tinylora.py]: wanna replace this with a slice assignment too?

File: adafruit_tinylora/adafruit_tinylora.py [hunk 2]
```diff
@@ -235,15 +219,15 @@ def send_data(self, data, data_length, frame_counter, timeout=2):
         # set length of LoRa packet
         lora_pkt_len = 9
         # load encrypted data into lora_pkt
-        for i in range(0, data_length):
+        for i in range(data_length):
```
Comment (ladyada) [inline adafruit_tinylora/adafruit_tinylora.py]: ditto, slice

File: adafruit_tinylora/adafruit_tinylora.py [hunk 3]
```diff
@@ -235,15 +219,15 @@ def send_data(self, data, data_length, frame_counter, timeout=2):
...
             lora_pkt[lora_pkt_len + i] = enc_data[i]
         # recalculate packet length
         lora_pkt_len = lora_pkt_len + data_length
         # Calculate MIC
         mic = bytearray(4)
         mic = aes.calculate_mic(lora_pkt, lora_pkt_len, mic)
         # load mic in package
-        for i in range(0, 4):
+        for i in range(4):
```
Comment (ladyada) [inline adafruit_tinylora/adafruit_tinylora.py]: slice

File: adafruit_tinylora/adafruit_tinylora_encryption.py [hunk 4]
```diff
@@ -150,7 +133,7 @@ def _aes_encrypt(self, data, key):
             for row in range(4):
                 state[col][row] = data[row + (col << 2)]
         # copy key to round_key
-        round_key = bytearray(16)
+        round_key = bytearray(key)
         for i in range(16):
```
Comment (ladyada) [inline adafruit_tinylora/adafruit_tinylora_encryption.py]: remove the assignemnt loop

File: adafruit_tinylora/adafruit_tinylora_encryption.py [hunk 5]
```diff
@@ -345,27 +328,21 @@ def _mic_generate_keys(self, key_1, key_2):
...
         # shift k1 left 1b
         self._shift_left(key_1)
         # check if msb is 1
-        if msb_key == 1:
-            key_1[15] = key_1[15] ^ 0x87
+        if msb_key:
+            key_1[15] ^= 0x87
         # perform gen_key on key_2
         # copy key_1 to key_2
         for i in range(16):
```
Comment (ladyada) [inline adafruit_tinylora/adafruit_tinylora_encryption.py]: slice

**Review by ladyada — CHANGES_REQUESTED** [review]

File: adafruit_tinylora/adafruit_tinylora_encryption.py (hunk 5 above)
Comment (brentru) [reply adafruit_tinylora/adafruit_tinylora_encryption.py]: done

File: adafruit_tinylora/adafruit_tinylora.py [hunk 6]
```diff
@@ -219,16 +218,14 @@ def send_data(self, data, data_length, frame_counter, timeout=2):
         # set length of LoRa packet
         lora_pkt_len = 9
         # load encrypted data into lora_pkt
```
Comment (ladyada) [inline adafruit_tinylora/adafruit_tinylora.py:220]: slice is wrong

File: adafruit_tinylora/adafruit_tinylora.py [hunk 7]
```diff
@@ -199,8 +199,7 @@ def send_data(self, data, data_length, frame_counter, timeout=2):
         enc_data = bytearray(data_length)
         lora_pkt = bytearray(64)
         # copy bytearray into bytearray for encryption
```
Comment (ladyada) [inline adafruit_tinylora/adafruit_tinylora.py:201]: second slice shoukd also be datalen

File: adafruit_tinylora/adafruit_tinylora.py [hunk 8]
```diff
@@ -219,16 +218,14 @@ def send_data(self, data, data_length, frame_counter, timeout=2):
         # set length of LoRa packet
         lora_pkt_len = 9
         # load encrypted data into lora_pkt
-        for i in range(data_length):
-            lora_pkt[lora_pkt_len + i] = enc_data[i]
+        lora_pkt[lora_pkt_len:data_length] = enc_data[0:4]
         # recalculate packet length
         lora_pkt_len = lora_pkt_len + data_length
```
Comment (ladyada) [inline adafruit_tinylora/adafruit_tinylora.py]: change to +=

**Review by ladyada — CHANGES_REQUESTED** [review]

**Comment by jerryneedell** [comment]
ran tinylora_simpletest.py  with this PR on an itsybitsy_m4_express and a Raspberry pi 3b+  -- worked fine. Fixed the issue with the Raspberry Pi import of ttn_usa.py  -- ran just before last commit -- will repeat ;-)

repeated test with last commit -- no issues.

**Comment by ladyada** [comment]
thanks @jerryneedell -im just tryin' to squish as much as possible so a feather m0 is happy :)

File: adafruit_tinylora/adafruit_tinylora.py (hunk 6 above)
Comment (ladyada) [reply adafruit_tinylora/adafruit_tinylora.py:220]: yay

File: adafruit_tinylora/adafruit_tinylora.py (hunk 8 above)
Comment (brentru) [reply adafruit_tinylora/adafruit_tinylora.py]: Changed in latest commit.

File: adafruit_tinylora/adafruit_tinylora_encryption.py (hunk 4 above)
Comment (brentru) [reply adafruit_tinylora/adafruit_tinylora_encryption.py]: Removed

File: adafruit_tinylora/adafruit_tinylora.py (hunk 3 above)
Comment (brentru) [reply adafruit_tinylora/adafruit_tinylora.py]: slice'd

File: adafruit_tinylora/adafruit_tinylora.py (hunk 2 above)
Comment (brentru) [reply adafruit_tinylora/adafruit_tinylora.py]: slice'd

**Comment by jerryneedell** [comment]
Understood!  -- just ran it on a feather_m0_express -- used .mpy's -- ran OK.

File: adafruit_tinylora/adafruit_tinylora.py [hunk 9]
```diff
@@ -235,16 +218,14 @@ def send_data(self, data, data_length, frame_counter, timeout=2):
         # set length of LoRa packet
         lora_pkt_len = 9
         # load encrypted data into lora_pkt
-        for i in range(0, data_length):
-            lora_pkt[lora_pkt_len + i] = enc_data[i]
+        lora_pkt[lora_pkt_len:data_length] = enc_data[0:data_length]
```
Comment (ladyada) [inline adafruit_tinylora/adafruit_tinylora.py]: fix 
 lora_pkt[lora_pkt_len:lora_pkt_len+data_length] = enc_data[0:data_length]

**Review by ladyada — CHANGES_REQUESTED** [review]

**Comment by ladyada** [comment]
one typo i saw, otherwise i think good, but you should do some testing also!

File: adafruit_tinylora/adafruit_tinylora.py (hunk 9 above)
Comment (jerryneedell) [reply adafruit_tinylora/adafruit_tinylora.py]: oops -- just realized my feather_m0 test was using older version -- updated to last commit - and it does work -- seeing this, I'm not sure why...

Comment (ladyada) [reply adafruit_tinylora/adafruit_tinylora.py]: i think that circuitpython/python is a little 'loose' on slice assignments. its best to be Correct, tho?

Comment (brentru) [reply adafruit_tinylora/adafruit_tinylora.py]: fixed.

**Comment by brentru** [comment]
Short-term test looked good. I'm loading the newest commit onto a Feather M0 LoRa and a M4 w/a Si7021 to run overnight for testing .

**Comment by ladyada** [comment]
looks good to me so far. thanks for the fixes. lets see if the mpy size improves

**Comment by ladyada** [comment]
@brentru wanna do a bump minor release?

**Comment by ladyada** [comment]
looks like 700 bytes of savings, just in mpy size. not bad!

**Pull request merged**